lib/phoenix/channel/server.ex

```diff
@@ -314,11 +339,33 @@ defmodule Phoenix.Channel.Server do
         %{topic: topic} = socket
       ) do
     start = System.monotonic_time()
-    result = socket.channel.handle_in(event, payload, put_in(socket.ref, ref))
-    duration = System.monotonic_time() - start
-    metadata = %{ref: ref, event: event, params: payload, socket: socket}
-    :telemetry.execute([:phoenix, :channel_handled_in], %{duration: duration}, metadata)
-    handle_in(result)
+    metadata = %{event: event, params: payload, socket: socket, ref: ref}
+    measurements = %{system_time: System.system_time()}
+    :telemetry.execute([:phoenix, :channel, :handle_in, :start], measurements, metadata)
+
+    try do
+      result = socket.channel.handle_in(event, payload, put_in(socket.ref, ref))
+      state = handle_in(result)
+
+      measurements = %{duration: System.monotonic_time() - start}
+
+      :telemetry.execute(
+        [:phoenix, :channel, :handle_in, :stop],
+        measurements,
+        Map.merge(metadata, %{socket: get_socket(state), result: result})
+      )
+
+      :telemetry.execute([:phoenix, :channel_handled_in], measurements, metadata)
+
+      state
```

Comment:
This isn't state of the GenServer, but the return value of handle_info when joining the channel is also called "state". I'm happy to change it if someone has a better idea.